### PR TITLE
Migrating to OSBA

### PR DESCRIPTION
**Pull request by zhongyi-zhang**
This PR includes the guidance and scripts for migrating MASB to [OSBA](https://github.com/Azure/open-service-broker-azure).

Supported services to migrate:
* azure-sqldb
* azure-rediscache
* azure-storage
* azure-servicebus
* azure-eventhubs

### CODE REVIEW

File: docs/migration/migrate_to_new_azure_service_broker.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+2. Get the `manifest.yml`(filled for `cf push`) of both meta-azure-service-broker and open-service-broker-azure.
+
+3. Run:
+
+```
+$ chmod +x migrate_to_new_azure_service_broker.sh
+$ ./migrate_to_new_azure_service_broker.sh <path-to-masb-manifest> <path-to-asb-manifest> (<path-to-new-sql-server-list>)
+```
+
+The broker database migration succeeds if the script ends with no error message.
```
Comment (bingosummer) [inline docs/migration/migrate_to_new_azure_service_broker.md]: How to deal with the instances which failed to migrate?

File: docs/migration/migrate_to_new_azure_service_broker.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+  Please check following credentials in the meta-azure-service-broker manifest:
+    * SUBSCRIPTION_ID
+    * TENANT_ID
+    * CLIENT_ID
+    * CLIENT_SECRET
+  
+* If you hit `Broken instance record` issue:
+  
+  Please check if the record is valid in meta-azure-service-broker database by running `select * from instances where instanceID='<instanceID>'`. Delete it and retry if it is invalid.
```
Comment (bingosummer) [inline docs/migration/migrate_to_new_azure_service_broker.md]: Is there a better way than just deleting it?

File: docs/migration/migrate_to_new_azure_service_broker.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+```
+$ cf update-service-broker <service-broker-name> <admin-to-new-broker> <password-to-new-broker> <url-to-new-broker>
+```
+
+The `<service-broker-name>` can be checked by `cf service-brokers` if you forget which name meta-azure-service-broker was registered with.
+
+## Notes (IMPORTANT)
+
+* The bound apps possiblely need some code changes, as the format of credentials delivered by the broker have some changes. Please refer to [open-service-broker-azure docs](https://github.com/Azure/open-service-broker-azure/tree/master/docs/modules) and [meta-azure-service-broker docs](https://github.com/Azure/meta-azure-service-broker/tree/master/docs) to see the differences module-by-module in detail. Note that the credentials just change in new bindings. Credentials in existing bindings won't change.
```
Comment (bingosummer) [inline docs/migration/migrate_to_new_azure_service_broker.md]: Should users `unbind` and `bind` the service instance?

Comment (zhongyi-zhang) [reply docs/migration/migrate_to_new_azure_service_broker.md]: If their apps have already done the code change for the new credentials provided by OSBA, they should. Or, they shouldn't. Will add this.

File: docs/migration/migrate_to_new_azure_service_broker.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
+# Migrate to new azure service broker
```
[the remark this answers is not in the document]
Comment (zhongyi-zhang) [reply docs/migration/migrate_to_new_azure_service_broker.md]: Thanks for the suggestions. Will refine.

File: docs/migration/migrate_to_new_azure_service_broker.md (hunk 1 above)
Comment (zhongyi-zhang) [reply docs/migration/migrate_to_new_azure_service_broker.md]: Will add a link to troubleshooting below.

File: docs/migration/migrate_to_new_azure_service_broker.md (hunk 3 above)
Comment (bingosummer) [reply docs/migration/migrate_to_new_azure_service_broker.md]: Did you add it?

File: docs/migration/README.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+  Please check following credentials in the MASB manifest:
+    * SUBSCRIPTION_ID
+    * TENANT_ID
+    * CLIENT_ID
+    * CLIENT_SECRET
+  
+* If you hit `Broken instance record` issue:
+  
+  Please confirm if it's an orphan record, the resouce on Azure in no longer existed, and the service instance is not in use. If yes, delete it in MASB database by running `delete from bindings where instanceID='<instanceID>'; delete from instances where instanceID='<instanceID>';` and retry the migration script.
```
Comment (bingosummer) [inline docs/migration/README.md]: Maybe users need to know how to login the MASB database.

File: docs/migration/migrate_to_new_azure_service_broker.md (hunk 3 above)
Comment (zhongyi-zhang) [reply docs/migration/migrate_to_new_azure_service_broker.md]: Added it now.

File: docs/migration/README.md (hunk 5 above)
Comment (zhongyi-zhang) [reply docs/migration/README.md]: I think there are too many tools to connect to a SQL DB. And the operators have the credentials. It should be very easy for them.

**Comment by zhongyi-zhang** [comment]
https://github.com/Azure/open-service-broker-azure/issues/150
Note it here for future update. The migration script in this PR will keep updated for the breaking changes in OSBA until we target a version to migrate.

**Comment by zhongyi-zhang** [comment]
Close as we would not support automatic migration.

**Pull request closed without merging**